src/vmm/src/builder.rs

```diff
@@ -148,26 +148,372 @@ impl std::convert::From<linux_loader::cmdline::Error> for StartMicrovmError {
+        )?;
+
+        Ok(())
+    }
+
+    /// Sets up the irqchip for a x86_64 microVM.
+    pub fn setup_interrupt_controller(vm: &mut Vm) -> Result<(), VmmError> {
+        vm.setup_irqchip().map_err(VmmError::Vm)
+    }
+
+    /// For x86_64 we need to create the interrupt controller before calling `KVM_CREATE_VCPUS`
+    /// while on aarch64 we need to do it the other way around.
+    pub fn create_vcpus(
+        vm: &mut Vm,
+        vcpu_count: u8,
+        exit_evt: &EventFd,
+    ) -> Result<Vec<Vcpu>, VmmError> {
+        let mut vcpus = Vec::with_capacity(vcpu_count as usize);
+        for cpu_idx in 0..vcpu_count {
+            let exit_evt = exit_evt.try_clone().map_err(VmmError::EventFd)?;
+            let vcpu = Vcpu::new(cpu_idx, vm, exit_evt).map_err(VmmError::VcpuCreate)?;
+            vcpus.push(vcpu);
+        }
+
+        Ok(vcpus)
+    }
+
+    pub(crate) fn load_kernel(
+        boot_config: &BootConfig,
+        guest_memory: &GuestMemoryMmap,
+    ) -> Result<GuestAddress, StartMicrovmError> {
+        let mut kernel_file = boot_config
+            .kernel_file
+            .try_clone()
+            .map_err(|err| StartMicrovmError::Internal(VmmError::KernelFile(err)))?;
+
+        let entry_addr = Loader::load::<std::fs::File, GuestMemoryMmap>(
+            guest_memory,
+            None,
+            &mut kernel_file,
+            Some(GuestAddress(crate::arch::get_kernel_start())),
+        )
+        .map_err(StartMicrovmError::KernelLoader)?;
+
+        Ok(entry_addr.kernel_load)
+    }
+
+    pub(crate) fn create_vmm_and_vcpus(
+        instance_info: &InstanceInfo,
+        guest_memory: GuestMemoryMmap,
+        uffd: Option<Uffd>,
+        vm_config: &VmConfig,
+        kvm_capabilities: Vec<KvmCapability>,
+    ) -> Result<(Vmm, Vec<Vcpu>), StartMicrovmError> {
+        let mut vmm = build_vmm(
+            instance_info,
+            guest_memory,
+            uffd,
+            vm_config,
+            kvm_capabilities,
+        )?;
+
+        let pio_device_manager = {
+            // Serial device setup.
+            let serial_device = setup_serial_device(std::io::stdin(), io::stdout())
+                .map_err(StartMicrovmError::Internal)?;
+
+            // x86_64 uses the i8042 reset event as the Vmm exit event.
+            let reset_evt = vmm
+                .vcpus_exit_evt
+                .try_clone()
+                .map_err(VmmError::EventFd)
+                .map_err(StartMicrovmError::Internal)?;
+
+            setup_interrupt_controller(&mut vmm.vm).map_err(StartMicrovmError::Internal)?;
+
+            // create pio dev manager with legacy devices
+            let pio_device_manager = {
+                // TODO Remove these unwraps.
+                let mut pio_dev_mgr = PortIODeviceManager::new(serial_device, reset_evt).unwrap();
+                pio_dev_mgr.register_devices(vmm.vm.fd()).unwrap();
+                pio_dev_mgr
+            };
+
+            pio_device_manager
+        };
```

Comment:
The `pio_device_manager` itself should be constructed where it was previously in the `build_vmm` and should not be `Optional` as there is no situation where x86 vmm does not have one. Only things you need to keep here are 2 lines:
```
setup_interrupt_controller(&mut vmm.vm).map_err(StartMicrovmError::Internal)?;
and 
pio_dev_mgr.register_devices(vmm.vm.fd()).unwrap();
```

Reply:
thanks for teaching me this 🙇🏻